### PR TITLE
Add a wptrun frontend to wptrunner.

### PR DESCRIPTION
**Pull request by jgraham**
This provides a convenient way for users to run tests in a browser
without having to grasp the many command line switches of wptrunner,
or install dependencies by hand. The syntax is basically

wptrun browser [tests]

e.g.

wptrun firefox dom/historical.html

Currently this is implemented for Firefox and Chrome, and will not
download the actual browser binary for you. For Firefox downloading
certutil is also a problem because there aren't convenient binary
releases. However we can still download the appropriate webdriver
implementations, firefox prefs, create a virtualenv, install required
wptrunner dependencies for the browser and generally make the process
of running tests substantially easier.

The keen-eyed with notice that this stole a lot of code from the
stability checker. The intent in the long term is cearly to re-merge
these codepaths so that we have a single implementation here.

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/5893)

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/231218675)

# Firefox (nightly) #
Testing web-platform-tests at revision a9ba65f70ea2875df8e2af6c68fe28ce6f7023ab
Using browser at version BuildID 20170511133312; SourceStamp 838652a84b76c273e084d0705f3f4f3be89520a8
Starting 10 test iterations
No tests run.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/231218681)

# Chrome (unstable) #
Testing web-platform-tests at revision a9ba65f70ea2875df8e2af6c68fe28ce6f7023ab

File: tools/browserutils/browser.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+            bits = "64" if uname[-1] == "x86_64" else "32"
+        elif platform == "mac":
+            bits = "64"
+        elif platform == "win":
+            bits = "32"
+
+        return "%s%s" % (platform, bits)
+
+    def install(self):
+        return None
```
Comment (jugglinmike) [inline tools/browserutils/browser.py]: I know we don't invoke this method today, but it seems like for the time being, we'd want to inherit the base class's implementation.

File: tools/browserutils/browser.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+
+    def install_webdriver(self, dest=None):
+        """Install latest Webdriver."""
+        if dest is None:
+            dest = os.pwd
+        latest = get("http://chromedriver.storage.googleapis.com/LATEST_RELEASE").text.strip()
+        url = "http://chromedriver.storage.googleapis.com/%s/chromedriver_%s.zip" % (latest,
+                                                                                     self.platform_string())
+        unzip(get(url).raw, dest)
+        path = os.path.join(dest, 'chromedriver')
```
Comment (jugglinmike) [inline tools/browserutils/browser.py]: Typo: single quotes

File: tools/browserutils/virtualenv.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+    @property
+    def pip_path(self):
+        path = find_executable("pip", self.bin_path)
+        if path is None:
+            raise ValueError("pip not found")
+        return path
+
+    def activate(self):
+        path = os.path.join(self.bin_path, 'activate_this.py')
```
Comment (jugglinmike) [inline tools/browserutils/virtualenv.py]: Typo: single quotes

**Review by jugglinmike — COMMENTED** [review]
I'm happy to see this developing; it's going to help a lot of people. I've tested it locally, and it seems to work as intended. Could we document its presence in the `README.md` file and including installation instructions?

File: .gitignore [hunk 4]
```diff
@@ -14,3 +14,4 @@ webdriver/.idea
 .vscode/
 .DS_Store
 *.rej
+_venv
```
Comment (gsnedders) [inline .gitignore]: Missing new line at the EOF

**Comment by jgraham** [comment]
Review status: 0 of 10 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[wptrun.py, line 95 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/5893#-KjsGG--j_RSQBjauOD--r1-95:-KjwGoT8PlLsltSdCKwT:bddu742) ([raw file](https://github.com/w3c/web-platform-tests/blob/1ddd05ce2154fc37ad2e538e4ff416b3e71b8aec/wptrun.py#L95)):*
<details><summary><i>Previously, jugglinmike wrote…</i></summary><blockquote>

I'd rather interpret "empty response" as invalid, not "yes".
</blockquote></details>

I quite strongly disagree. There's a long history of tools using empty response to mean "do whatever the sensible default is" e.g. apt.

---

*[tools/browserutils/browser.py, line 227 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/5893#-KjsGG-0URUMtzP9lvCO-r1-227:-KjwH7Hpur-zctYxk-tu:b-71pcef) ([raw file](https://github.com/w3c/web-platform-tests/blob/1ddd05ce2154fc37ad2e538e4ff416b3e71b8aec/tools/browserutils/browser.py#L227)):*
<details><summary><i>Previously, jugglinmike wrote…</i></summary><blockquote>

I know we don't invoke this method today, but it seems like for the time being, we'd want to inherit the base class's implementation.
</blockquote></details>

That doesn't work because the base class implementation is marked as abstract.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/5893)*

**Comment by bobholt** [comment]
Reviewed 5 of 11 files at r1, 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[tools/wptrun.py, line 95 at r1](https://reviewable.io:443/reviews/w3c/web-platform-tests/5893#-KjsGG--j_RSQBjauOD--r1-95:-KjwOZotT8RkX4eW-rqT:b-y39iih) ([raw file](https://github.com/w3c/web-platform-tests/blob/1ddd05ce2154fc37ad2e538e4ff416b3e71b8aec/tools/wptrun.py#L95)):*
> prompt_install
I see the appeal of forcing the user to make a choice, and if this were a new UI we creating, I'd want that to be the case. But since this is a command-line prompt, and it ends with `[Y/n]`?, in `apt` style, I'm fine with it as-is. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/5893)*

**Comment by bobholt** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/5893#-:-KjwQRFBv6Z4RPqQUkvF:bnfp4nl)*

**Comment by bobholt** [comment]
Reviewed 1 of 6 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[README.md, line 77 at r3](https://reviewable.io:443/reviews/w3c/web-platform-tests/5893#-KjxaZVoDMxSR8WPPrcY:-KjxaZVpchVsVMGWThkE:bpooaal) ([raw file](https://github.com/w3c/web-platform-tests/blob/4e76745df86c6f3f4e26028bc021c75713d83d77/README.md#L77)):*
> ```Markdown
> 
> Tests can be run automatically in a browser using the `wptrun` script
> in the root of the checkout. This requires the hosts file and OpenSSL
> ```

This is no longer in the root of the checkout, but in the `tools` directory, correct?

---

*[README.md, line 93 at r3](https://reviewable.io:443/reviews/w3c/web-platform-tests/5893#-KjxaAHK2rlUM5kZ9SQR:-KjxaAHK2rlUM5kZ9SQS:b-7z46u2) ([raw file](https://github.com/w3c/web-platform-tests/blob/4e76745df86c6f3f4e26028bc021c75713d83d77/README.md#L93)):*
> ```Markdown
> with `wptrun --help` and `wptrun --wptrunner-help`.
> 
> Not all dependnecies can be automatically installed; in paticular the
> ```

TYPOs: "dependencies", "particular"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/w3c/web-platform-tests/5893)*

File: tools/browserutils/browser.py [hunk 5]
```diff
@@ -0,0 +1,268 @@
+import logging
+import os
+import re
+import stat
+from abc import ABCMeta, abstractmethod
+from configparser import RawConfigParser
```
Comment (emanlove) [inline tools/browserutils/browser.py]: configparser is 3.x while ConfigParser is 2.x noting that the [web-platform-tests documentation](https://github.com/w3c/web-platform-tests#running-the-tests) states that "test environment requires Python 2.7+".

File: tools/browserutils/virtualenv.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            raise ValueError("virtualenv must be installed and on the PATH")
+
+    @property
+    def exists(self):
+        return os.path.isdir(self.path)
+
+    def create(self):
+        if os.path.exists(self.path):
+            shutil.rmtree(self.path)
+        call(self.virtualenv, "--no-download", self.path)
```
Comment (emanlove) [inline tools/browserutils/virtualenv.py]: The ```--no-download``` option was removed in [pip version 7.0.0](https://github.com/pypa/pip/blob/master/NEWS.rst#700-2015-05-21).

**Pull request merged**